DDDEastAnglia/Areas/Admin/Controllers/SponsorController.cs

```diff
@@ -0,0 +1,120 @@
+﻿using System;
+using System.Drawing;
+using System.Drawing.Imaging;
+using System.IO;
+using System.Linq;
+using System.Web;
+using System.Web.Mvc;
+using DDDEastAnglia.Areas.Admin.Models;
+using DDDEastAnglia.DataAccess;
+using DDDEastAnglia.Helpers;
+using DDDEastAnglia.Models;
+
+namespace DDDEastAnglia.Areas.Admin.Controllers
+{
+    public class SponsorController : Controller
+    {
+        private readonly ISponsorRepository sponsorRepository;
+        private readonly ISponsorSorter sponsorSorter;
+
+        public SponsorController(ISponsorRepository sponsorRepository, ISponsorSorter sponsorSorter)
+        {
+            if (sponsorRepository == null)
+            {
+                throw new ArgumentNullException("sponsorRepository");
+            }
+
+            if (sponsorRepository == null)
+            {
+                throw new ArgumentNullException("sponsorRepository");
+            }
+
+            this.sponsorRepository = sponsorRepository;
+            this.sponsorSorter = sponsorSorter;
+        }
+
+        public ActionResult Index()
+        {
+            var sponsors = sponsorRepository.GetAllSponsors();
+            var sortedSponsors = sponsorSorter.Sort(sponsors)
+                                              .Select(CreateSponsorModel)
+                                              .ToList();
+            return View(sortedSponsors);
+        }
+
+        public ActionResult Create()
+        {
+            return View(new SponsorModel());
+        }
+
+        [HttpPost]
+        public ActionResult Create(SponsorModel sponsor)
+        {
+            sponsorRepository.AddSponsor(CreateSponsor(sponsor));
+            return RedirectToAction("Index");
+        }
+
+        public ActionResult Delete(int id)
+        {
+            var sponsor = sponsorRepository.GetSponsor(id);
+            return sponsor == null ? (ActionResult)HttpNotFound() : View(sponsor);
+        }
+
+        [HttpPost, ActionName("Delete")]
+        [ValidateAntiForgeryToken]
+        public ActionResult DeleteConfirmed(int id)
+        {
+            sponsorRepository.DeleteSponsor(id);
+            return RedirectToAction("Index");
+        }
+
+        public ActionResult Logo(int sponsorId)
+        {
+            var sponsor = sponsorRepository.GetSponsor(sponsorId);
+            return File(sponsor.Logo, "image/png");
+        }
+
+        private SponsorModel CreateSponsorModel(Sponsor sponsor)
```

Comment:
`SponsorModel` is more like a view model as it is used within the pages. `Sponsor` is the data access class. There is a subtle difference between the two. The view model doesn't contain the logo, but does contain the sponsor's id so that the logo can be loaded.

I agree that it's a bit smelly, but the two are needed so that the validation can be done on creation.

Any ideas on how to make it better?